File: upstream/ray-gauges/packages/gauge-gen/src/accounts/GaugeConfig.ts
import { PublicKey, Connection } from "@solana/web3.js"
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface GaugeConfigFields {
  /** Global token account hopper for RAY tokens to emit */
  rayHopper: PublicKey
  /** Total RAY emission per day */
  rayEmissionPerDay: BN
  /** Total votes across all gauges */
  totalVotes: BN
  /**
   * Global RAY reward index for all gauges
   * The value of RAY per-vote
   */
  index: types.NumberRawFields
  /** Time since last update */
  lastUpdatedTs: BN
  bump: Array<number>
}

export interface GaugeConfigJSON {
  /** Global token account hopper for RAY tokens to emit */
  rayHopper: string
  /** Total RAY emission per day */
  rayEmissionPerDay: string
  /** Total votes across all gauges */
  totalVotes: string
  /**
   * Global RAY reward index for all gauges
   * The value of RAY per-vote
   */
  index: types.NumberRawJSON
  /** Time since last update */
  lastUpdatedTs: string
  bump: Array<number>
}

/** Global config for all pool connected pool gauges */
export class GaugeConfig {
  /** Global token account hopper for RAY tokens to emit */
  readonly rayHopper: PublicKey
  /** Total RAY emission per day */
  readonly rayEmissionPerDay: BN
  /** Total votes across all gauges */
  readonly totalVotes: BN
  /**
   * Global RAY reward index for all gauges
   * The value of RAY per-vote
   */
  readonly index: types.NumberRaw
  /** Time since last update */
  readonly lastUpdatedTs: BN
  readonly bump: Array<number>

  static readonly discriminator = Buffer.from([
    162, 113, 190, 180, 118, 167, 38, 30,
  ])

  static readonly layout = borsh.struct([
    borsh.publicKey("rayHopper"),
    borsh.u64("rayEmissionPerDay"),
    borsh.u64("totalVotes"),
    types.NumberRaw.layout("index"),
    borsh.u64("lastUpdatedTs"),
    borsh.array(borsh.u8(), 1, "bump"),
  ])

  constructor(fields: GaugeConfigFields) {
    this.rayHopper = fields.rayHopper
    this.rayEmissionPerDay = fields.rayEmissionPerDay
    this.totalVotes = fields.totalVotes
    this.index = new types.NumberRaw({ ...fields.index })
    this.lastUpdatedTs = fields.lastUpdatedTs
    this.bump = fields.bump
  }

  static async fetch(
    c: Connection,
    address: PublicKey,
    programId: PublicKey = PROGRAM_ID
  ): Promise<GaugeConfig | null> {
    const info = await c.getAccountInfo(address)

    if (info === null) {
      return null
    }
    if (!info.owner.equals(programId)) {
      throw new Error("account doesn't belong to this program")
    }

    return this.decode(info.data)
  }

  static async fetchMultiple(
    c: Connection,
    addresses: PublicKey[],
    programId: PublicKey = PROGRAM_ID
  ): Promise<Array<GaugeConfig | null>> {
    const infos = await c.getMultipleAccountsInfo(addresses)

    return infos.map((info) => {
      if (info === null) {
        return null
      }
      if (!info.owner.equals(programId)) {
        throw new Error("account doesn't belong to this program")
      }

      return this.decode(info.data)
    })
  }

  static decode(data: Buffer): GaugeConfig {
    if (!data.slice(0, 8).equals(GaugeConfig.discriminator)) {
      throw new Error("invalid account discriminator")
    }

    const dec = GaugeConfig.layout.decode(data.slice(8))

    return new GaugeConfig({
      rayHopper: dec.rayHopper,
      rayEmissionPerDay: dec.rayEmissionPerDay,
      totalVotes: dec.totalVotes,
      index: types.NumberRaw.fromDecoded(dec.index),
      lastUpdatedTs: dec.lastUpdatedTs,
      bump: dec.bump,
    })
  }

  toJSON(): GaugeConfigJSON {
    return {
      rayHopper: this.rayHopper.toString(),
      rayEmissionPerDay: this.rayEmissionPerDay.toString(),
      totalVotes: this.totalVotes.toString(),
      index: this.index.toJSON(),
      lastUpdatedTs: this.lastUpdatedTs.toString(),
      bump: this.bump,
    }
  }

  static fromJSON(obj: GaugeConfigJSON): GaugeConfig {
    return new GaugeConfig({
      rayHopper: new PublicKey(obj.rayHopper),
      rayEmissionPerDay: new BN(obj.rayEmissionPerDay),
      totalVotes: new BN(obj.totalVotes),
      index: types.NumberRaw.fromJSON(obj.index),
      lastUpdatedTs: new BN(obj.lastUpdatedTs),
      bump: obj.bump,
    })
  }
}

File: upstream/ray-gauges/sol/programs/gauge/src/state/gauge_config.rs
use crate::{pda::GLOBAL_CONFIG_SEED, state::NumberRaw};
use anchor_lang::prelude::*;
use precise_number::Number;

const SECONDS_IN_DAY: u64 = 86_400;

/// Global config for all pool connected pool gauges
#[account]
pub struct GaugeConfig {
    /// Global token account hopper for RAY tokens to emit
    pub ray_hopper: Pubkey,

    /// Total RAY emission per day
    pub ray_emission_per_day: u64,

    /// Total votes across all gauges
    pub total_votes: u64,

    /// Global RAY reward index for all gauges
    /// The value of RAY per-vote
    pub index: NumberRaw,

    /// Time since last update
    pub last_updated_ts: u64,

    pub bump: [u8; 1],
}

impl GaugeConfig {
    pub const SIZE: usize =
        // discriminator
        8 +
        // ray_hopper
        32 +
        // ray_emission_per_day
        8 +
        // total_votes
        8 +
        // index
        NumberRaw::SIZE +
        // last_updated_ts
        8 +
        // bump
        1;

    pub fn seeds(&self) -> [&[u8]; 2] {
        [GLOBAL_CONFIG_SEED.as_bytes(), &self.bump]
    }

    /// Update the index to the latest value
    /// Each unit in the index is a "per-vote" share of RAY
    pub fn update_index(&mut self, now: u64) {
        assert!(
            now >= self.last_updated_ts,
            "cannot update with older timestamp"
        );

        let time_elapsed = now - self.last_updated_ts;

        if time_elapsed <= 0 {
            return;
        }

        // If there are no votes on the gauge, return early
        if self.total_votes == 0 {
            self.last_updated_ts = now;
            return;
        }

        // calculate the amount of RAY to emit for the duration
        let time_elapsed: u64 = time_elapsed.try_into().unwrap();
        // duration as a ratio of a day
        let duration_day = Number::from_ratio(time_elapsed.into(), SECONDS_IN_DAY.into());
        // RAY emission
        let ray_emission = Number::from_natural_u64(self.ray_emission_per_day) * duration_day;

        let total_shares = Number::from_natural_u64(self.total_votes);

        // RAY per vote share on gauges
        let ray_per_share = ray_emission / total_shares;

        let cur_index: Number = self.index.into();
        let new_index = cur_index + ray_per_share;

        self.index = new_index.into();
        self.last_updated_ts = now;
    }

    pub fn change_votes(&mut self, amount: i64) {
        let is_inc = amount > 0;
        let amount: u64 = amount.abs().try_into().unwrap();
        if is_inc {
            self.inc_votes(amount)
        } else {
            self.dec_votes(amount)
        }
    }

    fn inc_votes(&mut self, amount: u64) {
        self.total_votes = self
            .total_votes
            .checked_add(amount)
            .expect("total votes overflow");
    }

    fn dec_votes(&mut self, amount: u64) {
        self.total_votes = self
            .total_votes
            .checked_sub(amount)
            .expect("total votes underflow");
    }
}
